primary/src/grpc_server/validator.rs

```diff
@@ -2,57 +2,97 @@
 // SPDX-License-Identifier: Apache-2.0
 use std::time::Duration;
 
-use crate::{block_waiter::GetBlockResponse, BlockCommand};
+use crate::{block_waiter::GetBlockResponse, BlockCommand, BlockRemoverCommand};
 use tokio::{
-    sync::{mpsc::Sender, oneshot},
+    sync::{mpsc::channel, mpsc::Sender, oneshot},
     time::timeout,
 };
 use tonic::{Request, Response, Status};
 use types::{
-    BatchMessageProto, BlockError, CollectionRetrievalResult, GetCollectionsRequest,
-    GetCollectionsResponse, Validator,
+    BatchMessageProto, BlockError, CertificateDigest, CertificateDigestProto,
+    CollectionRetrievalResult, Empty, GetCollectionsRequest, GetCollectionsResponse,
+    RemoveCollectionsRequest, Validator,
 };
 
 #[derive(Debug)]
 pub struct NarwhalValidator {
     tx_get_block_commands: Sender<BlockCommand>,
     get_collections_timeout: Duration,
+    tx_block_removal_commands: Sender<BlockRemoverCommand>,
+    remove_collections_timeout: Duration,
 }
 
 impl NarwhalValidator {
     pub fn new(
         tx_get_block_commands: Sender<BlockCommand>,
         get_collections_timeout: Duration,
+        tx_block_removal_commands: Sender<BlockRemoverCommand>,
+        remove_collections_timeout: Duration,
     ) -> Self {
         Self {
             tx_get_block_commands,
             get_collections_timeout,
+            tx_block_removal_commands,
+            remove_collections_timeout,
         }
     }
 }
 
 #[tonic::async_trait]
 impl Validator for NarwhalValidator {
+    async fn remove_collections(
+        &self,
+        request: Request<RemoveCollectionsRequest>,
+    ) -> Result<Response<Empty>, Status> {
+        let collection_ids = request.into_inner().collection_ids;
+        let remove_collections_response = if !collection_ids.is_empty() {
+            let (tx_remove_block, mut rx_remove_block) = channel(1);
+            let ids = parse_certificate_digests(collection_ids)?;
+            self.tx_block_removal_commands
+                .send(BlockRemoverCommand::RemoveBlocks {
+                    ids,
+                    sender: tx_remove_block,
+                })
+                .await
+                .map_err(|err| Status::internal(format!("Send Error: {:?}", err)))?;
+            match timeout(self.remove_collections_timeout, rx_remove_block.recv()).await {
```

Comment:
You could match on `timeout(...).await.map_err(|_err| Status::internal("Timeout ..."))` to remove one level of right-nesting from this.

Reply:
Nice catch, thanks!